cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,240 @@
+    case TakeOptions::TO_NULL:
+      return UnpackValuesNullCount<TakeOptions::TO_NULL>(context, values, indices,
+                                                         builder);
+    case TakeOptions::UNSAFE:
+      return UnpackValuesNullCount<TakeOptions::UNSAFE>(context, values, indices,
+                                                        builder);
+    default:
+      return Status::NotImplemented("how did we get here");
+  }
+}
+
+template <typename IndexType>
+struct UnpackValues {
+  using IndexArrayRef = const typename TypeTraits<IndexType>::ArrayType&;
+
+  template <typename ValueType>
+  Status Visit(const ValueType&) {
+    using ValueArrayRef = const typename TypeTraits<ValueType>::ArrayType&;
+    using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+    IndexArrayRef indices = static_cast<IndexArrayRef>(*params_.indices);
+    ValueArrayRef values = static_cast<ValueArrayRef>(*params_.values);
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(params_.context->memory_pool(), values.type(), &builder));
+    RETURN_NOT_OK(builder->Reserve(indices.length()));
+    RETURN_NOT_OK(UnpackOutOfBoundsBehavior(params_.context, values, indices,
+                                            params_.options,
+                                            static_cast<OutBuilder*>(builder.get())));
+    return builder->Finish(params_.out);
+  }
+
+  Status Visit(const NullType& t) {
+    auto indices_length = params_.indices->length();
+    if (params_.options.out_of_bounds == TakeOptions::RAISE && indices_length != 0) {
+      auto indices = static_cast<IndexArrayRef>(*params_.indices).raw_values();
+      auto minmax = std::minmax_element(indices, indices + indices_length);
+      auto min = static_cast<int64_t>(*minmax.first);
+      auto max = static_cast<int64_t>(*minmax.second);
+      if (min < 0 || max >= params_.values->length()) {
+        return Status::Invalid("out of bounds index");
+      }
+    }
+    params_.out->reset(new NullArray(indices_length));
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) {
+    std::shared_ptr<Array> taken_indices;
+    {
+      // To take from a dictionary, apply the current kernel to the dictionary's
+      // indices. (Use UnpackValues<IndexType> since IndexType is already unpacked)
+      auto indices = static_cast<const DictionaryArray*>(params_.values.get())->indices();
+      TakeParameters params = params_;
+      params.values = indices;
+      params.out = &taken_indices;
+      UnpackValues<IndexType> unpack = {params};
+      RETURN_NOT_OK(VisitTypeInline(*t.index_type(), &unpack));
+    }
+    // create output dictionary from taken indices
+    return DictionaryArray::FromArrays(dictionary(t.index_type(), t.dictionary()),
+                                       taken_indices, params_.out);
+  }
+
+  Status Visit(const ExtensionType& t) {
```

Comment:
Please use a catch all instead of defining all those not-implemented overloads explicitly.

Reply:
Because I use a template [here](https://github.com/apache/arrow/pull/3880/files/3a1ef1285e591d71f6aca125c0662e6cc6c4eab0#diff-dbc961df861bbaa965256e5ad42236ceR141), `Visit(const DataType&)` would not catch these types.

I could use enable_if to exclude them from the template Visit, but that would also entail enumerating them and I'm not sure it would be more clear. What do you prefer?

Reply:
Fine, let's not bother then :-)